x-pack/plugins/apm/public/components/app/Main/ProvideBreadcrumbs.tsx

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Location } from 'history';
+import React from 'react';
+import {
+  matchPath,
+  RouteComponentProps,
+  RouteProps,
+  withRouter
+} from 'react-router-dom';
+
+type LocationMatch<T = any> = Pick<
+  RouteComponentProps<T>,
+  'location' | 'match'
+>;
+
+export type BreadcrumbFunction<T = any> = (props: LocationMatch<T>) => string;
```

Comment:
I'm unsure about all these generics. It seems like they always default to `any`.
1) can we avoid the default value alltogether, and actually pass in the params?
2) if we MUST have a default value, can you make it `unknown` instead of `any`?

Reply:
Yeah I was hoping I could figure out how to type the match params so each render method would have typed args. May not be possible though.

Reply:
I tried making all the generics required, and then passing an argument to `BreadcrumbRoute` in routeConfig.tsx:

```tsx
export const routes: Array<
  BreadcrumbRoute<{
    groupId: string;
    serviceName: string;
  }>
> = [
  {
    exact: true,
    path: '/',
    render: renderAsRedirectTo('/services'),
    breadcrumb: 'APM'
  },
...
```

It _almost_ works but still some kinks :/
I have a feeling it's possible, my TS skills are just not advanced enough.

Reply:
All this would do is that the params are automatically fed into `render`.

Reply:
Was more thinking the breadcrumb functions would be safer, but this is almost exactly the same as your table column problem and I'm not sure it's worth it here. Now that I need to redo a lot of this PR, I'll probably skip this and remove.

Reply:
Oh yeah! That's exactly what I mean (breadcrumb function) when I wrote `render`

Reply:
Got confused about which PR I need to redo actually lol, this one is fine